### PR TITLE
Merch v2

### LINKED ISSUE #377: Add options to decline t-shirt and/or polo
Also provide help text explaining how we do swag

### LINKED ISSUE #1482: Allow selecting alternate merch
Basically, implement this:

>This is my general plan or what I was thinking of: anyone who hasn’t gotten a polo needs one, so they shouldn’t be presented with this choice. If a member has had their polo marked as “picked up” in a previous semester, then they should be presented with a short survey. I was thinking put it maybe next to the resume block on the home screen. it should be a simple multiple choice like “Our records indicate you have already received a RoboJackets polo. Please select your preferred merchandise for this year’s spring dues: ” with an option for “Polo” and an option for the other thing (probably hoodie or whatever the PR chair decides to order that year). When they hit confirm, if they selected the other option that isn’t polo, maybe prompt them to enter their size for the merch. That’s all I was thinking. If the member doesn’t fill out the survey, it should default to polo. is this enough info?

Exact implementation details are TBD. The question should be prompted during the dues flow and the alternate(s) should be configurable.

### PR DESCRIPTION
**Pull request by JoshOldenburg**
Fixes #1482 and fixes #377.

### CODE REVIEW

File: resources/js/components/dues/DuesRequiredInfo.vue [hunk 1]
```diff
@@ -174,6 +196,46 @@ export default {
     localUser: function() {
       return this.user;
     },
+    selectedPackage: function() {
+      if (null === this.duesPackages) return null;
```
Comment (evan10s) [inline resources/js/components/dues/DuesRequiredInfo.vue:200]: You should enclose the body of the `if` statement in brackets so no one gets confused later

File: resources/js/components/dues/DuesRequiredInfo.vue [hunk 2]
```diff
@@ -174,6 +196,46 @@ export default {
...
+      return this.duesPackages.find(duespackage => duespackage.id == this.duesPackageChoice);
+    },
+    merchDependencyText: function() {
+      var base = 'The options depend on your dues term selection above';
+      if (!this.selectedPackage) return base + ', so please select that first.';
+      else return base + '.';
```
Comment (evan10s) [inline resources/js/components/dues/DuesRequiredInfo.vue:206]: brackets pls

**Review by evan10s — CHANGES_REQUESTED** [review]
I didn't run this locally, but there's nothing super glaring in the code

**Review by JoshOldenburg — COMMENTED** [review]
The pivot tables should not be editable in Nova. `Merchandise` detail should have the pivot columns defined.

File: app/Models/DuesTransactionMerchandise.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    protected $casts = [
+        'provided_at' => 'datetime',
+    ];
+
+    /**
+     * Indicates if the IDs are auto-incrementing.
+     *
+     * @var bool
+     */
+    public $incrementing = true;
```
Comment (kberzinch) [inline app/Models/DuesTransactionMerchandise.php:26]: 🤔

File: app/Models/DuesTransactionMerchandise.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        return $this->belongsTo(User::class, 'provided_by');
+    }
+
+    /**
+     * Returns the name field of the User in providedBy. This exists because Nova was not cooperating.
+     */
+    public function getProvidedByNameAttribute(): ?string
+    {
+        return optional($this->providedBy)->name;
+    }
```
Comment (kberzinch) [inline app/Models/DuesTransactionMerchandise.php:39]: This kills the link I assume? Can we add a tech debt item to try and fix this at some point?

File: app/Nova/DuesPackage.php [hunk 5]
```diff
@@ -126,7 +128,15 @@ public function fields(Request $request): array
...
             HasMany::make('Prevents Purchase Of', 'hasConflictWith', self::class),
 
-            new Panel('Swag', $this->swagFields()),
+            BelongsToMany::make('Merchandise')
+                ->fields(static function (): array {
+                    return [
+                        SelectOrCustom::make('Group')->options([
+                            'Fall' => 'Fall',
+                            'Spring' => 'Spring',
+                        ]),
```
Comment (kberzinch) [inline app/Nova/DuesPackage.php:137]: I feel like this should just be a normal Nova field.

File: database/migrations/2021_01_11_231447_migrate_swag_data.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                'fiscal_year_id' => $fy->id,
+            ]);
+
+            $fallPackageName = 'Fall '.$startingYear;
+            $springPackageName = 'Spring '.$endingYear;
+            $studentFullYearName = 'Full Year ('.$startingYear.'-'.$endingYear.')';
+
+            $fallPackage = $fy->packages()->where('name', $fallPackageName)->first();
+            $springPackage = $fy->packages()->where('name', $springPackageName)->first();
+            $studentFullYearPackage = $fy->packages()->where('name', $studentFullYearName)->first();
```
Comment (kberzinch) [inline database/migrations/2021_01_11_231447_migrate_swag_data.php:42]:
```suggestion
            $fallPackage = $fy->packages()->where('name', $fallPackageName)->sole();
            $springPackage = $fy->packages()->where('name', $springPackageName)->sole();
            $studentFullYearPackage = $fy->packages()->where('name', $studentFullYearName)->sole();
```

File: phpstan.neon [hunk 7]
```diff
@@ -64,3 +64,4 @@ parameters:
     - '#Return type \(Illuminate\\Database\\Eloquent\\Builder\|Illuminate\\Database\\Eloquent\\Relations\\HasMany\) of method App\\Nova\\Filters\\[a-zA-Z]+::apply\(\) should be covariant with return type \(Illuminate\\Database\\Eloquent\\Builder\) of method Laravel\\Nova\\Filters\\Filter::apply\(\)#'
     - '#Short ternary operator is not allowed\. Use null coalesce operator if applicable or consider using long ternary\.#'
     - '#Variable property access on App\\Models\\User\.#'
+    - '#Access to an undefined property App\\Models\\[a-zA-Z]+::\$pivot\.#'
```
Comment (kberzinch) [inline phpstan.neon:67]: Please sort this.

**Review by kberzinch — CHANGES_REQUESTED** [review]
Initial notes without click testing

File: app/Models/DuesTransactionMerchandise.php (hunk 3 above)
Comment (JoshOldenburg) [reply app/Models/DuesTransactionMerchandise.php:26]: Per Laravel docs this is required for pivot models if they have an autoincrementing ID column.

File: app/Models/DuesTransactionMerchandise.php (hunk 4 above)
Comment (JoshOldenburg) [reply app/Models/DuesTransactionMerchandise.php:39]: Yes, it does. I tried several different ways to get it to use the relationship on the pivot model to no avail.

File: database/migrations/2021_01_11_231447_migrate_swag_data.php (hunk 6 above)
Comment (JoshOldenburg) [reply database/migrations/2021_01_11_231447_migrate_swag_data.php:42]: I need this to return `null` if there is no matching package.

Comment (kberzinch) [reply database/migrations/2021_01_11_231447_migrate_swag_data.php:42]: I think all of those exist, but ok

**Review by evan10s — APPROVED** [review]

**Comment by JoshOldenburg** [comment]
Also fixes #377.

**Review by kberzinch — APPROVED** [review]

**Pull request merged**